File: code/java/src/onlinejudge/luogu/solved/P1782.java
package onlinejudge.luogu.solved;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class P1782 {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        String[] readTmp;

        int n, m, c;
        readTmp = bufferedReader.readLine().split(" ");
        n = Integer.parseInt(readTmp[0]);
        m = Integer.parseInt(readTmp[1]);
        c = Integer.parseInt(readTmp[2]);

        long[] dp = new long[c + 1];

        int count, vv, ww, dd, v, w;
        for (int i = 0; i < n; i++) {
            readTmp = bufferedReader.readLine().split(" ");
            vv = Integer.parseInt(readTmp[0]);
            ww = Integer.parseInt(readTmp[1]);
            dd = Integer.parseInt(readTmp[2]);
            count = 1;
            while (dd > count) {
                dd -= count;
                v = count * vv;
                w = count * ww;
                for (int j = c; j >= v; j--) {
                    dp[j] = Math.max(dp[j], dp[j - v] + w);
                }
                count <<= 1;
            }
            v = dd * vv;
            w = dd * ww;
            for (int j = c; j >= v; j--) {
                dp[j] = Math.max(dp[j], dp[j - v] + w);
            }
        }

        long aa, bb, cc;
        for (int i = 0; i < m; i++) {
            readTmp = bufferedReader.readLine().split(" ");
            aa = Long.parseLong(readTmp[0]);
            bb = Long.parseLong(readTmp[1]);
            cc = Long.parseLong(readTmp[2]);
            for (int j = c; j > -1; j--) {
                for (int x = 0; x <= j; x++) {
                    dp[j] = Math.max(dp[j], dp[j - x] + aa * x * x + bb * x + cc);
                }
            }
        }

        System.out.println(dp[c]);
    }
}

File: code/cpp/src/onlinejudge/luogu/solved/p1782.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    long i, j, n, m, c, vv, ww, dd, count, v, w, aa, bb, cc, x;
    cin >> n >> m >> c;

    vector<long> dp(1 + c, 0);

    for (i = 0; i < n; i++) {
        cin >> vv >> ww >> dd;
        count = 1;
        while (dd > count) {
            dd -= count;
            v = count * vv;
            w = count * ww;
            for (j = c; j >= v; j--) {
                dp[j] = max(dp[j], dp[j - v] + w);
            }
            count <<= 1;
        }
        v = dd * vv;
        w = dd * ww;
        for (j = c; j >= v; j--) {
            dp[j] = max(dp[j], dp[j - v] + w);
        }
    }

    for (i = 0; i < m; i++) {
        cin >> aa >> bb >> cc;
        for (j = c; j > -1; j--) {
            for (x = 0; x <= j; x++) {
                dp[j] = max(dp[j], dp[j - x] + aa * x * x + bb * x + cc);
            }
        }
    }

    cout << dp[c] << endl;

    return 0;
}
